### ISSUE TITLE
Add additional info to Yii2 assets guide 

### ISSUE THREAD

**Issue by PowerGamer1** [yiisoft/yii2#12017]
The guide http://www.yiiframework.com/doc-2.0/guide-structure-assets.html could benefit from covering the following topic: is it supported/recommended to modify `$js` / `$css` and `$jsOptions` / `$cssOptions` of already published assets bundle?

Example: MyWidget can be configured to require 1.js OR 2.js. Some page may only use instances of MyWidget configured for 1.js, other page may only use instances of MyWidget configured for 2.js and yet another page may use several instances of MyWidget configured for 1.js and at the same time several instances of MyWidget configured for 2.js.

Solution 1:
```PHP
class MyWidgetAsset1 extends \yii\web\AssetBundle
{
	public $sourcePath = 'somepath';
	public $js = ['1.js'];
}
class MyWidgetAsset2 extends \yii\web\AssetBundle
{
	public $sourcePath = 'somepath';
	public $js = ['2.js'];
}
class MyWidget extends \yii\base\Widget
{
	public function run()
	{
             if($this->config['type'] == 1)
                    MyWidgetAsset1::register($this->getView());
             else
                    MyWidgetAsset2::register($this->getView());
        }
}
```

Solution 2: (needs confirmation if supported/recommended by Yii2):
```PHP
class MyWidgetAsset extends \yii\web\AssetBundle
{
	public $sourcePath = 'somepath';
	public $js = [];
}
class MyWidget extends \yii\base\Widget
{
	public function run()
	{
             $bundle = MyWidgetAsset::register($this->getView());
             if($this->config['type'] == 1 && !in_array('1.js', $bundle->js))
                    $bundle->js[] = '1.js';
             else if(!in_array('2.js', $bundle->js))
                     $bundle->js[] = '2.js';
       }
}
```

**Comment by samdark** [comment]
It's recommended to create separate bundles in this case.

**Comment by PowerGamer1** [comment]
Nice, thanks for the clarification! 

Another unclear moment. Is it supported/recommended for two different asset bundles to require the same resource (.js for example) from the same ($sourcePath)?

For example, consider MyWidget example above implemented through Solution 1 (with separate bundles) but now MyWidget also always requires some base.js script. **That base.js script does not have any functionality that can be used on its own, it is assumed base.js should only be used in pair with 1.sj or 2.js.** Taking this into account, which variant is supported/recommended:

```PHP
class MyWidgetAsset1 extends \yii\web\AssetBundle
{
    public $sourcePath = 'somepath';
    public $js = ['base.js', '1.js'];
}
class MyWidgetAsset2 extends \yii\web\AssetBundle
{
    public $sourcePath = 'somepath';
    public $js = ['base.js', '2.js'];
}
```

```PHP
class MyWidgetBaseAsset extends \yii\web\AssetBundle
{
    public $sourcePath = 'somepath';
    public $js = ['base.js'];
}
class MyWidgetAsset1 extends \yii\web\AssetBundle
{
    public $sourcePath = 'somepath';
    public $js = ['1.js'];
    public $depends = [MyWidgetBaseAsset::class];
}
class MyWidgetAsset2 extends \yii\web\AssetBundle
{
    public $sourcePath = 'somepath';
    public $js = ['2.js'];
    public $depends = [MyWidgetBaseAsset::class];
}
```

P.S. The examples are taken from a real world case - a JavaScript charting library AmCharts, that has a base (common) script amcharts.js that is supposed to be paired with specific chart type scripts, for ex.:
for X-Y chart:
```HTML
<script src="amcharts/amcharts.js" type="text/javascript"></script>
<script src="amcharts/xy.js" type="text/javascript"></script>
```
for Pie chart:
```HTML
<script src="amcharts/amcharts.js" type="text/javascript"></script>
<script src="amcharts/pie.js" type="text/javascript"></script>
```
etc.

**Comment by samdark** [comment]
> Another unclear moment. Is it supported/recommended for two different asset bundles to require the same resource (.js for example) from the same ($sourcePath)?

It will be registered once if both bundles are used at the same time and isn't really recommended. It's much better to move `base.js` into separate bundle.

It does not matter much if you're combining your assets into a single file though.

**Comment by PowerGamer1** [comment]
Another problem, what if MyWidget supports translation into different languages and for each language has a separate .js file for inclusion (fr.js, ru.js etc.). There can be many such languages (for ex., AmCharts supports 32 languages) and creating a separate bundle for EVERY language seems unreasonable:
```PHP
class MyWidgetFrAsset extends \yii\web\AssetBundle
{
    public $sourcePath = 'somepath';
    public $js = ['fr.js'];
    public $depends = [MyWidgetBaseAsset::class];
}
class MyWidgetRuAsset extends \yii\web\AssetBundle
{
    public $sourcePath = 'somepath';
    public $js = ['ru.js'];
    public $depends = [MyWidgetBaseAsset::class];
}
// ...
```

Does Yii2 supports the reconfiguration of asset bundles (i.e. additional .js/.css files added/removed to bundle `$js` and `$css` arrays) AFTER the bundle is registered through `AssetBundle::register()` static method?

If yes then the solution could be:
```PHP
class MyWidgetLangAsset extends \yii\web\AssetBundle
{
    public $sourcePath = 'somepath';
    public $js = [];
    public $depends = [MyWidgetBaseAsset::class];
}
class MyWidget extends \yii\base\Widget
{
    public $language;
    public function run()
    {
         $assetBundle = MyWidgetLangAsset::register($this->getView());
         $langJs = $this->language . '.js';
         if(!in_array($langJs, $assetBundle->js))
             $assetBundle->js[] = $langJs;
    }
}
```

**Comment by klimov-paul** [comment]
> Does Yii2 supports the reconfiguration of asset bundles (i.e. additional .js/.css files added/removed to bundle $js and $css arrays) AFTER the bundle is registered through AssetBundle::register() static method?

See https://github.com/yiisoft/yii2-jui/blob/master/DatePickerLanguageAsset.php#L41

**Comment by PowerGamer1** [comment]
@klimov-paul
> See https://github.com/yiisoft/yii2-jui/blob/master/DatePickerLanguageAsset.php#L41

Very interesting.

The comment for `DatePickerLanguageAsset::$autoGenerate` says:
> If this is false, you should explicitly specify the language js files via [[js]].

Which means (correct me if I am wrong please) the code like the following in one of user views:
```PHP
echo \yii\jui\DatePicker::widget([/* ... */]);

$assetBundle = \yii\jui\DatePickerLanguageAsset::register($this);
$assetBundle->autoGenerate = false;
$assetBundle->js = ['ui/i18n/datepicker-ru.js', 'ui/i18n/datepicker-fr.js'];
```
But the code above is a prime example of bundle reconfiguration AFTER it was registered. So Yii2 supports such reconfiguration after all?

Next question is: what is the point of having `DatePickerLanguageAsset::$autoGenerate` at all? Why not implement it the way I showed for MyWidget/MyWidgetLangAsset above? Currently, the code:

```PHP
echo \yii\jui\DatePicker::widget([
	'model' => $model,
	'attribute' => 'x1',
	'language' => 'fr',
]);
echo \yii\jui\DatePicker::widget([
	'model' => $model,
	'attribute' => 'x2',
	'language' => 'ru',
]);
```
would produce two DatePickers with 'ru' language which is quite surprising and not obvious why.

**Comment by klimov-paul** [comment]
> But the code above is a prime example of bundle reconfiguration AFTER it was registered. So Yii2 supports such reconfiguration after all?

Of course it does - why not?
However any overriding at the level of `AssetManager` including 'asset compress' feature will fail to handle any bundle reconfiguration after it has been registered.

> Next question is: what is the point of having DatePickerLanguageAsset::$autoGenerate at all?

`autoGenerate` will compose the translation file using the general convension assuming it is present at the `jui` repository. You may want to use language, which is missing in standard repo or use localization different from the main one.

> would produce two DatePickers with 'ru' language which is quite surprising and not obvious why.

This is JS limitation - there is nothing we can do about it. You may report your concern to JQuery UI, to provide better localization support.

**Comment by PowerGamer1** [comment]
> However any overriding at the level of  AssetManager  including 'asset compress' feature will fail to handle any bundle reconfiguration after it has been registered.

Does this mean the current implementation of `DatePickerLanguageAsset` is incompatible with  'asset compress' feature (because the current interaction between `DatePiceker` and `DatePickerLanguageAsset` classes perform just that - reconfiguration of `DatePickerLanguageAsset` bundle after its registration with `AssetBundle::register()`)?

> `autoGenerate` will compose the translation file using the general convension assuming it is present at the  jui  repository. **You may want to use language, which is missing in standard repo or use localization different from the main one.**

Good point. The part in bold would really help in the description of `autoGenerate`.

> This is JS limitation - there is nothing we can do about it. You may report your concern to JQuery UI, to provide better localization support.

The following code will produce two DatePickers each with its own language on the same page without problems:
```PHP
echo \yii\jui\DatePicker::widget([
    'model' => $model,
    'attribute' => 'x1',
    'language' => 'fr',
]);
echo \yii\jui\DatePicker::widget([
    'model' => $model,
    'attribute' => 'x2',
    'language' => 'ru',
]);
$assetBundle = \yii\jui\DatePickerLanguageAsset::register($this);
$assetBundle->autoGenerate = false;
$assetBundle->js = ['ui/i18n/datepicker-ru.js', 'ui/i18n/datepicker-fr.js'];
```
So I don't think its a JUI limitation, the problem rather lies in the implementations of `DatePicker` and `DatePickerLanguageAsset`. User shouldn't have to reconfigure the `DatePickerLanguageAsset` bundle manually to have two DatePickers with different languages if the languages are supported by JUI.

**Comment by klimov-paul** [comment]
> Does this mean the current implementation of DatePickerLanguageAsset is incompatible with 'asset compress' feature

**Comment by klimov-paul** [comment]
> Does this mean the current implementation of DatePickerLanguageAsset is incompatible with 'asset compress' feature

Perhaps, since `sourcePath` will be overriden after compression. You may check it out to be sure.
However this asset bundle is a dinamic one and thus is seprated from any other. It is unlikely it will be used for asset compression.
Also there is a posibility the config for `asset/compress` command can be composed with this bundle covered and still be compressed in correct way.

> Good point. The part in bold would really help in the description of autoGenerate.

Feel free to open a PR for this. My opinion: it is obvious.

> the problem rather lies in the implementations of DatePicker and DatePickerLanguageAsset

If you have an idea for the improvement - feel free to open a PR.
I do not consider this is a crucial, since it is unlikely you will use several date pickers with different languages at the same page.

**Comment by PowerGamer1** [comment]
Just did a test to see if `DatePickerLanguageAsset` can be reconfigured at the level of AssetManager (following http://www.yiiframework.com/doc-2.0/guide-structure-assets.html#customizing-asset-bundles guide which menations nothing about `autoGenerate`):

at app config:
```PHP
		'assetManager' => [
			'bundles' => [
				yii\jui\DatePickerLanguageAsset::class => [
					'js' => [
						'https://ajax.googleapis.com/ajax/libs/jqueryui/1.11.4/i18n/datepicker-fr.js',
					]
				],
			],
		],
```

at view:
```PHP
echo \yii\jui\DatePicker::widget([
	'model' => $model,
	'attribute' => 'x1',
	'language' => 'fr',
]);
```

And as expected it didn't work properly: the page included:
```PHP
<script src="https://ajax.googleapis.com/ajax/libs/jqueryui/1.11.4/i18n/datepicker-fr.js"></script>
<script src="/assets/5877e356/ui/i18n/datepicker-fr.js"></script>
```
instead of only including .js file from https://ajax.googleapis.com. Setting `autoGenerate` to `false` does produce correct result though:

```PHP
		'assetManager' => [
			'bundles' => [
				yii\jui\DatePickerLanguageAsset::class => [
					'autoGenerate' => false,
					'js' => [
						'https://ajax.googleapis.com/ajax/libs/jqueryui/1.11.4/i18n/datepicker-fr.js',
					]
				],
			],
		],
```

Yet, if my app supports language switching, how am I supposed to reconfigure `DatePickerLanguageAsset` bundle at the level of AssetManager to include language .js file from `'https://ajax.googleapis.com` for current `\Yii::$app->language`?

**Comment by PowerGamer1** [comment]
@klimov-paul 
> You may want to use language, which is missing in standard repo or use localization different from the default one.

Lets say I want to use existing JUI languages while replacing ru.js and adding X.js. I tried to do that using `autoGenerate` and don't really see how it can be possible without one of the following:
a) putting my overridden (ru.js) and new (X.js) language files inside `@vendor` folder (`@vendor\bower\jquery-ui\ui\`).
b) copying all the existing content of `@vendor\bower\jquery-ui\ui\` into web-accessible folder and putting my files (ru.js and X.js) there too.

Both variants introduce a maintainance nightmare (ideally I would want to configure `DatePickerLanguageAsset` **at AssetManager level** in a way that it would pick standard JUI languages from `@vendor\bower\jquery-ui\ui\` and my overridden and new .js files from some other totally unrelated folder). How it was supposed to be done when `autoGenerate` was introduced?

**Comment by klimov-paul** [comment]
> Lets say I want to use existing JUI languages while replacing ru.js and adding X.js. I tried to do that using autoGenerate and don't really see how it can be possible without one of the following

You should not use `autoGenerate` in this case. That is what I was saying to you here:
https://github.com/yiisoft/yii2/issues/12017#issuecomment-236143996

**Comment by PowerGamer1** [comment]
> You should not use  autoGenerate  in this case. That is what I was saying to you here:
#12017 (comment)

Can you please provide an example how it could be done (since your reply implied that a way to do it is supposed to exist) with or without `autoGenerate`?

**Comment by klimov-paul** [comment]
Resolved by commit 65c60a0c2477117d07a64f45af8fa6c58f3edf62
